libtenzir/builtins/operators/chart.cpp

```diff
@@ -162,13 +179,34 @@ class chart_operator final : public crtp_operator<chart_operator> {
     // Cache attribute-enriched schemas, to avoid the potentially expensive
     // operation of building a list of attributes by visiting `cfg_` for every
     // iteration
+    auto limit = uint64_t{0};
+    {
+      auto l = limit_as_number(cfg_);
+      TENZIR_ASSERT(l);
+      limit = *l;
+    }
+    auto remaining = limit;
     std::unordered_map<type, type> enriched_schemas_cache{};
     previous_values_type previous_values{};
     for (auto&& slice : input) {
       if (slice.rows() == 0) {
         co_yield {};
         continue;
       }
+      if (remaining == 0) {
+        co_yield {};
+        continue;
+      }
+      if (slice.rows() > remaining) {
+        slice = subslice(slice, 0, remaining);
+        diagnostic::warning("chart reached event limit of `{}`", limit)
+          .hint("You can use `--limit `value` to change this limit")
+          .primary(loc_)
+          .emit(ctrl.diagnostics());
+        remaining = 0;
+      } else {
+        remaining -= slice.rows();
+      }
```

Comment:
I think this has a bug: If you get slices that exactly add up to the limit, then you won't get a warning.

Also, I think we can always `co_return` after emitting the warning.

Reply:
Ah right, that totally makes sense. But you can make it so that you limit to N events, but actually fetch at least N+1 (if there are more than N), then emit a warning if N is exceeded and do a `co_return` straight after.